[llvm-objdump] Implement decoding auxiliary header for xcoff with llvm-objdump --private-headers

diggerlin opened the pull request:
Implement decoding auxiliary header of XCOFF object file with llvm-objdump --private-headers

Commit:
implement decoding auxiliary header for xcoff with llvm-objdump --private-headers

Commit:
implement decoding auxiliary header for xcoff with llvm-objdump --private-headers

diggerlin committed:
minor change

diggerlin committed:
add more FileCheck options

jh7370 commented on llvm/test/tools/llvm-objdump/XCOFF/private-header-auxiliary.test:
```diff
@@ -0,0 +1,128 @@
+# CHECK32-NEXT:Maxium alignment of .data:          0x3
+# CHECK32-NEXT:Module type;                        0x0
+# CHECK32-NEXT:CPU type of objects:                0x1
+# CHECK32-NEXT:(Reserved):                         0x0
```
Does a Reserved field really need printing?

diggerlin replied:
thank,  I deleted it.

jh7370 commented on llvm/test/tools/llvm-objdump/XCOFF/private-header-auxiliary.test:
```diff
@@ -0,0 +1,128 @@
+# RUN: llvm-objdump --private-headers %t1 | \
+# RUN:   FileCheck %s --check-prefix=CHECK32 --match-full-lines --strict-whitespace
+# RUN: cp %t1 %t1_err1
+# RUN: python -c "with open(r'%t1_err1', 'r+b') as input: input.seek(17); input.write(b'\x45')"
+# RUN: python -c "with open(r'%t1_err1', 'r+b') as input: input.seek(4); input.write(b'\x00')"
```
Is there a reason these two lines can't be folded together into one python statement? You may also wish to look at using split-file, to allow you to have a proper python script, rather than these inline snippets.

diggerlin replied:
> you may also wish to look at using split-file, to allow you to have a proper python script, rather than these inline snippets.
This a one/two lines python, I do not think it need a split-file.

jh7370 commented on llvm/tools/llvm-objdump/XCOFFDump.cpp:
```diff
@@ -67,6 +75,179 @@ void XCOFFDumper::printStrHex(StringRef Name, StringRef Str, uint64_t Value) {
+        ArrayRef<uint8_t>(reinterpret_cast<const uint8_t *>(&AuxHeader) +
+                              PartialFieldOffset,
+                          AuxSize - PartialFieldOffset));
+  } else if (sizeof(AuxHeader) < AuxSize)
```
Please review the coding standards here: https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements. Note in particular the need for uniformity in an if/else-if chain.

jh7370 commented on llvm/tools/llvm-objdump/XCOFFDump.cpp:
```diff
@@ -67,6 +75,179 @@ void XCOFFDumper::printStrHex(StringRef Name, StringRef Str, uint64_t Value) {
+    Dumper->reportUniqueWarning(Twine("only partial field for ") +
+                                PartialFieldName + " at offset (" +
+                                Twine(PartialFieldOffset) + ")");
+    Dumper->printBinary(
```
You are presumably aware that this will print to stdout, whereas the warning will be printed to stderr, so they may not line up together?

jh7370 commented on llvm/tools/llvm-objdump/XCOFFDump.cpp:
```diff
@@ -67,6 +75,179 @@ void XCOFFDumper::printStrHex(StringRef Name, StringRef Str, uint64_t Value) {
 
+void XCOFFDumper::printBinary(StringRef Name, ArrayRef<uint8_t> B) {
+  unsigned OrgWidth = getWidth();
+  setWidth(0);
+  outs() << formatName(Name) << " (" << format_bytes(B) << ")\n";
+  setWidth(OrgWidth);
```
There's a lot of `setWidth` etc calls going on here, but I don't see any testing of the formatting for this case? I.e. if you removed one or other of these, I'm not sure you'd get a test failure?

diggerlin replied:
> There's a lot of `setWidth` etc calls going on here, but I don't see any testing of the formatting for this case? I.e. if you removed one or other of these, I'm not sure you'd get a test failure?

I think following test it.

`# RUN: llvm-objdump --private-headers %t1_extra 2>&1 | FileCheck %s --check-prefix=EXTRA --match-full-lines --strict-whitespace`

jh7370 replied:
> if you strong suggest to remove it, I can remove it.

I was actually suggesting that you should have more testing, with my suggestion being that you could tell if you had sufficient testing by removing a line of code and seeing if a test failed. However, I think your explanation makes sense, thanks, so no need to change things.

jh7370 commented on llvm/tools/llvm-objdump/XCOFFDump.cpp:
```diff
@@ -67,6 +75,179 @@ void XCOFFDumper::printStrHex(StringRef Name, StringRef Str, uint64_t Value) {
+}
+
+void XCOFFDumper::printAuxiliaryHeader(
+    const XCOFFAuxiliaryHeader64 *AuxHeader) {
```
I think you should look into how you can share the majority of this function with the 32-bit case. There's otherwise a lot of duplication.

diggerlin replied:
we prefer to print the order of  the fields of it as.  the  code can not be shared in the implement, but it output will give the user the information of the order of the fields.

jh7370 replied:
Okay, fair enough (though I do wonder how useful knowing the field order actually is for most users).

Could you at least put the `outs() << "\n---Auxiliary Header:\n";` line into a trivial helper function (e.g. `printAuxHeaderTitle`), so that it's not repeated in both versions? That way if you decide to change how this title is printed, you only have to change it in one place.

diggerlin commented on llvm/test/tools/llvm-objdump/XCOFF/private-header-auxiliary.test:
```diff
@@ -0,0 +1,128 @@
+## Test that `the llvm-objdump --private-headers` print out the auxiliary header of XCOFF object file.
```
yes, I think so.

diggerlin committed:
address comment

diggerlin committed:
reorder the run commands and add comments for test case

diggerlin commented:
My apologies for the delay in addressing the comments.

diggerlin commented:
> > My apologies for the delay in addressing the comments.
> 
> No need to apologise, as I now have to apologise as I'm about to go away on PTO and won't be able to look at this for 3-4 weeks at least. If it's urgent, I don't mind others confirming my comments have been addressed and this being merged and I can follow up review. Otherwise, I'll take a look once back.

Thank you for letting me know! No problem at all. I’m happy to wait for your review when you’re back

diggerlin commented:
> According to the pre-merge checks, this test is failing on Linux. Please review and fix.

jh7370 commented on llvm/test/tools/llvm-objdump/XCOFF/private-header-auxiliary.test:
```diff
@@ -0,0 +1,102 @@
+
+## Test how llvm-objdump behaves when the auxiliary header of the XCOFF object file contains a partial field.
+# RUN: cp %t1 %t1_err1
+# RUN: python -c "with open(r'%t1_err1', 'r+b') as input: input.seek(17); input.write(b'\x45'); input.seek(4); input.write(b'\x00')"
```
The problem is likely that you're calling `python` without using the `%python` substitution, here and below. Take a look at carefully at how other tests use `python` (e.g. https://github.com/llvm/llvm-project/blob/ba5676cf91f91bbddfacae06c036cf79af0f2088/llvm/test/tools/llvm-readobj/basic.test#L29)

diggerlin replied:
thanks for pointing out the error.

diggerlin committed:
change python to %python

jh7370 commented on llvm/tools/llvm-objdump/XCOFFDump.cpp:
```diff
@@ -67,6 +88,181 @@ void XCOFFDumper::printStrHex(StringRef Name, StringRef Str, uint64_t Value) {
+  ptrdiff_t Offset = reinterpret_cast<const char *>(&Member) -
+                     reinterpret_cast<const char *>(AuxHeader);
+  if (Offset + sizeof(Member) <= AuxSize)
+    Style == Hex ? printHex(MemberName, Member)
+                 : printNumber(MemberName, Member);
```
Whilst it's not incorrect, I can't say I've ever seen a ternary used where you just want a function call and don't want the return value. I'd suggest just using a plain `if` here.

jh7370 commented on llvm/tools/llvm-objdump/XCOFFDump.cpp:
```diff
@@ -67,6 +88,181 @@ void XCOFFDumper::printStrHex(StringRef Name, StringRef Str, uint64_t Value) {
+                                       const char *&PartialFieldName) {
+  ptrdiff_t Offset = reinterpret_cast<const char *>(&Member) -
+                     reinterpret_cast<const char *>(AuxHeader);
+  if (Offset + sizeof(Member) <= AuxSize)
```
I think the recommended coding style is that this should have braces, given the `else` does.

jh7370 reviewed (commented):
Sorry for the delay in reviewing again. A few more small comments and then I think we're done.

diggerlin commented:
> Sorry for the delay in reviewing again. A few more small comments and then I think we're done.

No need to apologize; thank you very much for your professional feedback!

diggerlin committed:
address comment

jh7370 reviewed (approved):
LGTM.
